File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
    var counts = {}
    var result = 0
    var answer = 0

    function add(key) {
        counts[key] = (counts[key] || 0) + 1;
        if (counts[key] == number[want.indexOf(key)]) {
            result += 1
        }
    }

    function reduce(key) {
        counts[key] = counts[key] - 1;
        if (counts[key] == number[want.indexOf(key)] - 1) {
            result -= 1
        }
    }
    
    for (let i = 0; i < 10; i++){
        add(discount[i])
    }
    if (result == want.length) {
        answer += 1
    }

    for (let i = 10; i < discount.length; i++) {
        add_target = discount[i]
        reduce_target = discount[i-10]
        add(add_target)
        reduce(reduce_target)
        if (result == want.length) {
            answer += 1
        }
    }

    
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <map>
#include <unordered_set>
#include <iostream>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    
    map<string, int> required_number;
    map<string, int> cur_number;
    unordered_set<string> want_set;
    int cur_counter = 0;
    int max_counter = number.size();
    
    for (string fruit: want) {
        want_set.insert(fruit);
    }
    
    for (int i = 0; i < want.size(); i++) {
        required_number[want[i]] = number[i];
    }
            
    for (int i = 0; i < 10; i++) {
        string cur_fruit = discount[i];
        cur_number[cur_fruit] += 1;
        if (cur_number[cur_fruit] == required_number[cur_fruit]) {
            cur_counter += 1;
        } 
    }
    if (cur_counter == max_counter) {
        answer += 1;
    }
    
    
    for (int i = 0; i < discount.size() - 10; i++) {
        string pop_fruit = discount[i];
        string add_fruit = discount[i+10];
        
        
        
        cur_number[pop_fruit] -= 1;
        cur_number[add_fruit] += 1;
        
        if (add_fruit == pop_fruit) {
            if (cur_counter == max_counter) {
                answer += 1;
            }
            continue;
        }
        
        if (want_set.count(pop_fruit) == 1 && cur_number[pop_fruit] == required_number[pop_fruit] - 1) {
            cur_counter -= 1;
        } 
        
        if (want_set.count(add_fruit) == 1 && cur_number[add_fruit] == required_number[add_fruit]) {
            cur_counter += 1;
        }
        
        if (cur_counter == max_counter) {
            answer += 1;
        }
    }
    
        
    
    
    

    return answer;
}